---
title: "ClusterJudge"
author: "Adrian Pasculescu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

references:
- author:
  - family: Gibbons
    given: F.D
  - family: Roth
    given: F.P.
  container-title: Genome Research
  id: GibRoth2002
  issued:
    month: October
    volume: 12
    year: 2002
  publisher: Genome Research
  title: 'Judging the Quality of Gene Expression-Based Clustering Methods Using Gene Annotation'
- author:
  - family: Cho
    given: R.C. et all
  container-title: Molecular Cell
  id: Cho1998
  issued:
    month: July
    volume: 2
    year: 1998
  publisher: Science Direct
  title: 'A genome-wide transcriptional analysis of the mitotic cell cycle'
- author:
  - family: Tavazoie
    given: S. et all
  container-title: Nature Genetics
  id: Taivasoie1999
  issued:
    month: July
    volume: 22
    year: 1999
  publisher: Nature Publishing Group
  title: 'Systematic determination of genetic network architecture'
- author:
  - family: Cover
    given: T.M. et all
  container-title: book
  id: Cover1991
  issued:
    month: 
    volume: (ed. D.L. Schilling)
    year: 1991
  publisher: Wiley-Interscience
  title: 'Elements of information theory'
---

## Introduction
Clustering can be judged based on external information.
For example when the clustered entities posses attributes, the total mutual information between
the clustering and each of the important attributes of entities might be helpful.
A `god` clustering should have a pronounced decrease of the total mutual information 
when the of number of random swaps of entities between clusters is increased [@GibRoth2002].

For example Gene Ontology attributes can help in assessing gene clustering resulting from 
genomics or proteomics experiments.

The following simple steps are needed for judging a clustering:
1. Obtain the clusters as a a structure containing entitities and their cluster ids
2. Obtain the entitity-attribute associations and Verify the entitity-attribute associations for consistency
3. In most of the cases one can consolidate (reduce) the number of attributes
by using a set of _uncertainty_ levels 
4. 'Judge' the original clustreing versus some random changes applied on the original one


## Obtaining the clustering

In most of the cases the clustering is obtained from experimental data after 
applying some statistical methods such as k-means or cutting a hierarchical clustering at a predifined 'height'.

The following example extracts the Yeast cell cycle clustering of genes from [@Cho1998] [@Taivasoie1999]
```{r}
library('yeastExpData')
data(ccyclered)
head(ccyclered)
```

The clustering should not contain ambiguites: i.e. one entity (gene) should be assigned to just one cluster
(and no NA !)
For this package the clusters must be organized as a named vector:
```{r}
clusters <- ccyclered$Cluster
###  convert from Gene names to the new standard of Saccharomyces Genome Database (SGD) gene ids
ccyclered$SGDID <- sub('^S','S00',ccyclered$SGDID)
names(clusters) <- ccyclered$SGDID
str(clusters)
```

Note that we modified the names of the genes to match the names of the entities in 
the Gene Ontology (see next section).


## Obtaining Entity Attribute associations


In the case of genomics or proteomics data, one source of attributes is the Gene Ontology (http://www.geneontology.org/).
For Yeast Gene Ontology attributes can be downloaded also from the [Saccharomyces Genome Database (SGD)]
(http://www.yeastgenome.org/)
In ClusterJudge R pacakge there is an already downloaded mapping of Yeast genes to attributes: 

```{r, fig.show='hold', echo=TRUE}
data(Yeast.GO.assocs);
str(Yeast.GO.assocs);
head(Yeast.GO.assocs);
validate.association(Yeast.GO.assocs)
```

A function to download a `fresh` copy of the mapping is also available as a function *download.Yeast.GO.mapping()* 

_validate.association()_ function quickly verifies if the input structure has two columns, 
if there are no NAs or NULLs and if thre are no duplicated associations.

## Consolidating Entity-Attribute associations
In some cases the attributes might not bring additional useful information.
 * For example an attribute can be too specific: assigned to one or two genes (can be seen as simply 
a new name of a gene).
 * In some other cases two or more attributes can be very correlated: appear on almost the same set of genes.

For the first case a consolidated association will ignore those attributes that belong to very few genes 
(*min.entities.per.attr*) 

```{r, fig.width=6, fig.height=4}
entities.attribute.stats(Yeast.GO.assocs) ### shows number of entities per attribute distribution
Yeast.GO.assocs.cons1 <- consolidate.entity.attribute(
       entity.attribute = Yeast.GO.assocs
     , min.entities.per.attr =3  ### keep only attributes associated to 3 or more entities
     , mut.inf=FALSE
     )

dim(Yeast.GO.assocs)
dim(Yeast.GO.assocs.cons1) ### shows reduction in the number of associations
``` 
For the second case we use the pre-calculated mutual information [@Cover1991] between attributes and impose some maximum of 
_uncertainty_ levels. 
Where uncertainty is defined as mutual information divided by the maximum of mutual information.

```{r, fig.width=6, fig.height=4}
data(mi.GO.Yeast)
Yeast.GO.assocs.cons <- consolidate.entity.attribute(
     entity.attribute = Yeast.GO.assocs
   , min.entities.per.attr =3
   , mut.inf=mi.GO.Yeast   ### use precalculated mutual information
   , U.limit = c(0.01, 0.001) ### calculate consolidated association for these uncertainty levels
   ) ### shows distribution of the number of pairs of attributes by Uncertainty 
str(Yeast.GO.assocs.cons)                                                   
```

The lower the Uncertainty limit the more attributes are ignored and the faster the judging 
of clusters will be performed.


## Calculating the mutual information 
In case the entity-attributes associations are downloaded from other organisms or other sources
mutual information can be calculated using this function: *attribute.mut.inf* :

```{r}
data(Yeast.GO.assocs) 
### because it takes time, we use a small sampled subset of associations
entity.attribute.sampled <- Yeast.GO.assocs[sample(1:nrow(Yeast.GO.assocs),500),]
mi.GO.Yeast.sampled <- attribute.mut.inf(
    entity.attribute = entity.attribute.sampled
  , show.progress    = FALSE  ## for this small example do not print progress
  )  
str(mi.GO.Yeast.sampled)
```

The mutual information of the pair of attributes A,B is defined as 
  mi(A,b) = H(A)+H(B) - H(cbind(A,B))
where H is the entropy.


## Judging the clustering for selected `uncertainty` levels
We use the _clusters_ named vector prepared above and one of the consolidated entity-attribute
associations (for example the one at 0.001 level of Uncertainty.
The _ClusterJudge_ will swap at random more and more entities between clusters and calculate
and plot the new total mutual information. If the total mutual information decreases by
increasing the number of random swaps we can colclude that there is a good agreement between
the clustering of experimental data and the ensmble of information provided by the
entity-attribute assocoations.

```{r, fig.width=6, fig.height=4}
mi.by.swaps<-clusterJudge(
    clusters = clusters
  , entity.attribute=Yeast.GO.assocs.cons[["0.001"]]
  , plot.notes='Yeast clusters judged at uncertainty level 0.001 - Ref: Tavazoie S,& all 
`Systematic determination of genetic network architecture. Nat Genet. 1999`')  
```

The most interesting cases might be when very reliable experimental data and sound clustering
contradict information accumulated in time into entity-attribute associations. These cases
might suggest novel discoveries.

## References


    
    

  